host/xtest/pkcs11_1000.c

```diff
@@ -2565,3 +2565,277 @@ static void xtest_pkcs11_test_1011(ADBG_Case_t *c)
+	ADBG_EXPECT_EQUAL(c, g_value, obj_value, sizeof(obj_value));
+	for (i = sizeof(obj_value); i < sizeof(g_value); i++)
+		if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, g_value[i], ==, 0xCC))
+			break;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	/*
+	 * Sub test: value with larger buffer
+	 */
+	Do_ADBG_BeginSubCase(c, "Get Attribute - value with larger buffer");
+	memset(g_value, 0xCC, sizeof(g_value));
+
+	rv = C_GetAttributeValue(session, obj_hdl, get_attr_template_vl,
+				 ARRAY_SIZE(get_attr_template_vl));
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, get_attr_template_vl[0].ulValueLen, ==, sizeof(obj_value));
+	ADBG_EXPECT_EQUAL(c, g_value, obj_value, sizeof(obj_value));
+	for (i = sizeof(obj_value); i < sizeof(g_value); i++)
+		if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, g_value[i], ==, 0xCC))
+			break;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	/*
+	 * Sub test: bool + value with exact size + class + bool
+	 */
+	Do_ADBG_BeginSubCase(c, "Get Attribute - bool + value with exact size + class + bool");
+	memset(g_value, 0xCC, sizeof(g_value));
+	g_token = CK_TRUE;
+	g_private = CK_TRUE;
+	g_class = ~0;
+
+	rv = C_GetAttributeValue(session, obj_hdl, get_attr_template_bvecb,
+				 ARRAY_SIZE(get_attr_template_bvecb));
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, get_attr_template_bvecb[1].ulValueLen,
+				     ==, sizeof(obj_value));
+	ADBG_EXPECT_EQUAL(c, g_value, obj_value, sizeof(obj_value));
+	for (i = sizeof(obj_value); i < sizeof(g_value); i++)
+		if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, g_value[i], ==, 0xCC))
+			break;
+
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, g_class, ==, CKO_DATA);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, g_token, ==, CK_FALSE);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, g_private, ==, CK_FALSE);
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	/*
+	 * Sub test: bool + value with larger buffer + class + bool
+	 */
+	Do_ADBG_BeginSubCase(c, "Get Attribute - bool + value with larger buffer + class + bool");
+	memset(g_value, 0xCC, sizeof(g_value));
+	g_token = CK_TRUE;
+	g_private = CK_TRUE;
+	g_class = ~0;
+
+	rv = C_GetAttributeValue(session, obj_hdl, get_attr_template_bvlcb,
+				 ARRAY_SIZE(get_attr_template_bvlcb));
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, get_attr_template_bvlcb[1].ulValueLen,
+				     ==, sizeof(obj_value));
+	ADBG_EXPECT_EQUAL(c, g_value, obj_value, sizeof(obj_value));
+	for (i = sizeof(obj_value); i < sizeof(g_value); i++)
+		if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, g_value[i], ==, 0xCC))
+			break;
+
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, g_class, ==, CKO_DATA);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, g_token, ==, CK_FALSE);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, g_private, ==, CK_FALSE);
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	/*
+	 * Sub test: allowed mechanism list
+	 */
+	Do_ADBG_BeginSubCase(c, "Get Attribute - allowed mechanism list");
+	memset(g_mecha_list, 0, sizeof(g_mecha_list));
```

Comment:
prefer reset to 0xFF as 0 is a valid mechanism id.

Reply:
OK

Reply:
This seems to backfire because of:
https://github.com/OP-TEE/optee_client/blob/master/libckteec/src/serialize_ck.c#L146

It is actually incorrect to use PKCS11_UNDEFINED_ID in here as that is not in the standard (and we are not yet in PKCS11 TA translated side) -- so removing the if clause would seem to be the right thing to do.

Also it is problematic for 64 bit systems. 

I do have experimental "sign extend for ~0UL" in testing for de-serializing CK_ULONG's to have some translation support for 64 bits systems. As we are not filling the extra memory then it should not matter but if there would have been value coming back then this would be a problem in here too.

Also the comparison for validity at later spot is a bit harder as there is no official define for it in PKCS11.

One could also use 0xCC or so as a value in here and then change the end comparison to be byte based. Might do that byte comparison anyway as that is compliant with both 32 & 64 bit systems. Value we can change to 0xFF if wanted to be so.

Reply:
Updated the PR with 0xCC